### ISSUE TITLE
Allow Github CI actions to run on custom local hardware

### ISSUE THREAD

**Issue by axel-h** [seL4/seL4#509]
- Can we make Github CI actions run on custom local hardware also? 
- In a repo fork, what needs to be adapted that this is tested locally (both hardware and AWS proofs)
- Assume there are multiple test farms, how could this be handled?

**Comment by lsf37** [comment]
Yes, that would be very nice to do. I haven't finished yet setting up the UNSW one for the rest of the repos and wanted to do that before thinking more about how to distribute it, but we can start discussing this now.

The current architecture for hardware tests is that there is
 * the GitHub CI action that compiles images
 * followed by an action that downloads those images and selects the correct board for it, then uses the UNSW machine queue commands to
     * request a lock for that board
     * issue a `run` command that uploads the image(s) and runs the test, getting output on the console and a log file in the end
     * release the lock
     * parse the log file for success/error

The machine queue is currently a set of shell scripts in [seL4/machine_queue](https://github.com/seL4/machine_queue) (currently still private, to be published, committers have read access at least atm) plus another set of python scripts to handle the actual power reset/boot/console for each board (also still private).

The latter is specific to the setup at UNSW. The former could be expanded to a reasonable larger interface. It's not published yet, because it currently requires `ssh` access to UNSW machines to use it, and @wom-bat  wanted to implement a web-hook solution for submitting tests first so that we can move away from that. There is also no reason that this has to be implemented as shell scripts, that's just historically grown (although it works fine for what it needs to so far).

In terms of how that interface could work, I would basically not change anything for the command line of `mq.sh`, but for each board, the scripts should know which test server to contact, instead of always going to UNSW. So if an organisation wants to add a new board to the test set, it would add the parameters that we have now, plus which server (or which submission URL if that web-hook interface is better in the end).

`mq.sh` currently calls an interface `reboot` with a bunch of options that actually runs the test. That is the bit each org would implement for themselves (or copy the UNSW setup if that is convenient). As mentioned, this is currently over `ssh`, so maybe not ideal, although if set up correctly that doesn't have to be a problem.

What I'm not clear on yet for the web-hook idea is how we would get live logs from the test back, which is much nicer than just a logfile in the end, because you can see if things are hanging or otherwise failing and then cancel tests.

Does this make some kind of sense? I haven't thought through this very deeply yet.

**Comment by lsf37** [comment]
Logging here that we will likely need to adjust some parts of #618 to enable forks to do their own hardware tests.

**Comment by kent-mcleod** [comment]
Closing this while it's not being actively worked on.
